sklearn/cluster/_hierarchical.pyx

```diff
@@ -332,3 +332,120 @@ cdef class WeightedEdge:
+################################################################################
+# Efficient labelling/conversion of MSTs to single linkage hierarchies
+
+cdef class UnionFind(object):
+
+    cdef ITYPE_t next_label
+    cdef ITYPE_t[:] parent
+    cdef ITYPE_t[:] size
+
+    def __init__(self, N):
+        self.parent = -1 * np.ones(2 * N - 1, dtype=ITYPE, order='C')
+        self.next_label = N
+        self.size = np.hstack((np.ones(N, dtype=ITYPE),
+                               np.zeros(N - 1, dtype=ITYPE)))
+
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef void union(self, ITYPE_t m, ITYPE_t n):
+        self.parent[m] = self.next_label
+        self.parent[n] = self.next_label
+        self.size[self.next_label] = self.size[m] + self.size[n]
+        self.next_label += 1
+
+        return
+
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef ITYPE_t fast_find(self, ITYPE_t n):
+        cdef ITYPE_t p
+        p = n
+        # find the highest node in the linkage graph so far
+        while self.parent[n] != -1:
+            n = self.parent[n]
+        # provide a shortcut up to the highest node
+        while self.parent[p] != n:
+            p, self.parent[p] = self.parent[p], n
+        return n
+
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+cpdef np.ndarray[DTYPE_t, ndim=2] _single_linkage_label(
+    np.ndarray[DTYPE_t, ndim=2] L):
+    """
+    Convert an linkage array or MST to a tree by labelling clusters at merges.
+    This is done by using a Union find structure to keep track of merges
+    efficiently. This is the private version of the function that assumes that
+    ``L`` has been properly validated. See ``single_linkage_label`` for the
+    user facing version of this function.
+
+    Parameters
+    ----------
+    L: array of shape (n_samples - 1, 3)
+        The linkage array or MST where each row specifies two samples
+        to be merged and a distance or weight at which the merge occurs. This
+         array is assumed to be sorted by the distance/weight.
+
+    Returns
+    -------
+    A tree in the format used by scipy.cluster.hierarchy.
+    """
+
+    cdef np.ndarray[DTYPE_t, ndim=2] result_arr
+    cdef DTYPE_t[:, ::1] result
+
+    cdef ITYPE_t left, left_cluster, right, right_cluster, index
+    cdef DTYPE_t delta
+
+    result_arr = np.zeros((L.shape[0], 4), dtype=DTYPE)
+    result = result_arr
+    U = UnionFind(L.shape[0] + 1)
+
+    for index in range(L.shape[0]):
+
+        left = <ITYPE_t> L[index, 0]
+        right = <ITYPE_t> L[index, 1]
+        delta = L[index, 2]
+
+        left_cluster = U.fast_find(left)
+        right_cluster = U.fast_find(right)
+
+        result[index][0] = left_cluster
+        result[index][1] = right_cluster
+        result[index][2] = delta
+        result[index][3] = U.size[left_cluster] + U.size[right_cluster]
+
+        U.union(left_cluster, right_cluster)
+
+    return result_arr
+
+
+def single_linkage_label(L):
```

Comment:
This seems very much like a pure Python function (it has not typing information). Any reason to have it in a Cython file?

Reply:
I was keeping it together with the private cythonized function above -- this is the future-proof wrapper in case future users wish to use the routine safely (i.e. with appropriate checks). It is not currently called at all. I am certainly happy to move it, but it is not clear to me what the appropriate place is at this time.